File: rb/SpacexRest_sdk.rb
# SpacexRest SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'SpacexRest_types'


class SpacexRestSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = SpacexRestUtility.new
    @_utility = utility

    config = SpacexRestConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = SpacexRestHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = SpacexRestHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, SpacexRestFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    SpacexRestUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = SpacexRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = SpacexRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = SpacexRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = SpacexRestSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue SpacexRestError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = SpacexRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = SpacexRestHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Idiomatic facade: client.capsule.list / client.capsule.load({ "id" => ... })
  def capsule
    require_relative 'entity/capsule_entity'
    @capsule ||= CapsuleEntity.new(self, nil)
  end

  # Deprecated: use client.capsule instead.
  def Capsule(data = nil)
    require_relative 'entity/capsule_entity'
    CapsuleEntity.new(self, data)
  end


  # Idiomatic facade: client.core.list / client.core.load({ "id" => ... })
  def core
    require_relative 'entity/core_entity'
    @core ||= CoreEntity.new(self, nil)
  end

  # Deprecated: use client.core instead.
  def Core(data = nil)
    require_relative 'entity/core_entity'
    CoreEntity.new(self, data)
  end


  # Idiomatic facade: client.crew.list / client.crew.load({ "id" => ... })
  def crew
    require_relative 'entity/crew_entity'
    @crew ||= CrewEntity.new(self, nil)
  end

  # Deprecated: use client.crew instead.
  def Crew(data = nil)
    require_relative 'entity/crew_entity'
    CrewEntity.new(self, data)
  end


  # Idiomatic facade: client.landpad.list / client.landpad.load({ "id" => ... })
  def landpad
    require_relative 'entity/landpad_entity'
    @landpad ||= LandpadEntity.new(self, nil)
  end

  # Deprecated: use client.landpad instead.
  def Landpad(data = nil)
    require_relative 'entity/landpad_entity'
    LandpadEntity.new(self, data)
  end


  # Idiomatic facade: client.launch.list / client.launch.load({ "id" => ... })
  def launch
    require_relative 'entity/launch_entity'
    @launch ||= LaunchEntity.new(self, nil)
  end

  # Deprecated: use client.launch instead.
  def Launch(data = nil)
    require_relative 'entity/launch_entity'
    LaunchEntity.new(self, data)
  end


  # Idiomatic facade: client.launchpad.list / client.launchpad.load({ "id" => ... })
  def launchpad
    require_relative 'entity/launchpad_entity'
    @launchpad ||= LaunchpadEntity.new(self, nil)
  end

  # Deprecated: use client.launchpad instead.
  def Launchpad(data = nil)
    require_relative 'entity/launchpad_entity'
    LaunchpadEntity.new(self, data)
  end


  # Idiomatic facade: client.payload.list / client.payload.load({ "id" => ... })
  def payload
    require_relative 'entity/payload_entity'
    @payload ||= PayloadEntity.new(self, nil)
  end

  # Deprecated: use client.payload instead.
  def Payload(data = nil)
    require_relative 'entity/payload_entity'
    PayloadEntity.new(self, data)
  end


  # Idiomatic facade: client.roadster.list / client.roadster.load({ "id" => ... })
  def roadster
    require_relative 'entity/roadster_entity'
    @roadster ||= RoadsterEntity.new(self, nil)
  end

  # Deprecated: use client.roadster instead.
  def Roadster(data = nil)
    require_relative 'entity/roadster_entity'
    RoadsterEntity.new(self, data)
  end


  # Idiomatic facade: client.rocket.list / client.rocket.load({ "id" => ... })
  def rocket
    require_relative 'entity/rocket_entity'
    @rocket ||= RocketEntity.new(self, nil)
  end

  # Deprecated: use client.rocket instead.
  def Rocket(data = nil)
    require_relative 'entity/rocket_entity'
    RocketEntity.new(self, data)
  end


  # Idiomatic facade: client.ship.list / client.ship.load({ "id" => ... })
  def ship
    require_relative 'entity/ship_entity'
    @ship ||= ShipEntity.new(self, nil)
  end

  # Deprecated: use client.ship instead.
  def Ship(data = nil)
    require_relative 'entity/ship_entity'
    ShipEntity.new(self, data)
  end


  # Idiomatic facade: client.starlink.list / client.starlink.load({ "id" => ... })
  def starlink
    require_relative 'entity/starlink_entity'
    @starlink ||= StarlinkEntity.new(self, nil)
  end

  # Deprecated: use client.starlink instead.
  def Starlink(data = nil)
    require_relative 'entity/starlink_entity'
    StarlinkEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = SpacexRestSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/spacexrest_sdk.py
# SpacexRest SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import SpacexRestUtility
from core.spec import SpacexRestSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import SpacexRestBaseFeature
from features import _make_feature


class SpacexRestSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = SpacexRestUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return SpacexRestUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = SpacexRestSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Capsule(self, data=None) -> "CapsuleEntity":
        """Entity factory: client.Capsule().list({}) / client.Capsule().load({"id": ...})."""
        from entity.capsule_entity import CapsuleEntity
        return CapsuleEntity(self, data)


    def Core(self, data=None) -> "CoreEntity":
        """Entity factory: client.Core().list({}) / client.Core().load({"id": ...})."""
        from entity.core_entity import CoreEntity
        return CoreEntity(self, data)


    def Crew(self, data=None) -> "CrewEntity":
        """Entity factory: client.Crew().list({}) / client.Crew().load({"id": ...})."""
        from entity.crew_entity import CrewEntity
        return CrewEntity(self, data)


    def Landpad(self, data=None) -> "LandpadEntity":
        """Entity factory: client.Landpad().list({}) / client.Landpad().load({"id": ...})."""
        from entity.landpad_entity import LandpadEntity
        return LandpadEntity(self, data)


    def Launch(self, data=None) -> "LaunchEntity":
        """Entity factory: client.Launch().list({}) / client.Launch().load({"id": ...})."""
        from entity.launch_entity import LaunchEntity
        return LaunchEntity(self, data)


    def Launchpad(self, data=None) -> "LaunchpadEntity":
        """Entity factory: client.Launchpad().list({}) / client.Launchpad().load({"id": ...})."""
        from entity.launchpad_entity import LaunchpadEntity
        return LaunchpadEntity(self, data)


    def Payload(self, data=None) -> "PayloadEntity":
        """Entity factory: client.Payload().list({}) / client.Payload().load({"id": ...})."""
        from entity.payload_entity import PayloadEntity
        return PayloadEntity(self, data)


    def Roadster(self, data=None) -> "RoadsterEntity":
        """Entity factory: client.Roadster().list({}) / client.Roadster().load({"id": ...})."""
        from entity.roadster_entity import RoadsterEntity
        return RoadsterEntity(self, data)


    def Rocket(self, data=None) -> "RocketEntity":
        """Entity factory: client.Rocket().list({}) / client.Rocket().load({"id": ...})."""
        from entity.rocket_entity import RocketEntity
        return RocketEntity(self, data)


    def Ship(self, data=None) -> "ShipEntity":
        """Entity factory: client.Ship().list({}) / client.Ship().load({"id": ...})."""
        from entity.ship_entity import ShipEntity
        return ShipEntity(self, data)


    def Starlink(self, data=None) -> "StarlinkEntity":
        """Entity factory: client.Starlink().list({}) / client.Starlink().load({"id": ...})."""
        from entity.starlink_entity import StarlinkEntity
        return StarlinkEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "SpacexRestSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.capsule_entity import CapsuleEntity
    from entity.core_entity import CoreEntity
    from entity.crew_entity import CrewEntity
    from entity.landpad_entity import LandpadEntity
    from entity.launch_entity import LaunchEntity
    from entity.launchpad_entity import LaunchpadEntity
    from entity.payload_entity import PayloadEntity
    from entity.roadster_entity import RoadsterEntity
    from entity.rocket_entity import RocketEntity
    from entity.ship_entity import ShipEntity
    from entity.starlink_entity import StarlinkEntity
